File: lecture_010(RecursionwithArraylists)/l_001_getsubsequence.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

vector<string> getss(string str) {
    if(str.length()==0) {
        vector<string> bres;
        bres.push_back("");
        return bres;
    }
    char ch = str[0];    //a
    string ros = str.substr(1);  //bc
    vector<string> rres = getss(ros);

    vector<string> myans;
    for(string rstr: rres) {
        myans.push_back("-"+ rstr);
    }
    for(string rstr: rres) {
        myans.push_back(ch+ rstr);
    }
    return myans;
}
int main() {
    string str;
    //cin>>str;
    getline(cin,str);
    vector<string> result = getss(str);
    cout<<"[";
    for(int i=0; i<result.size()-1; i++) {
        cout<<result[i]<<", ";
    }
    cout<<result[result.size()-1]<<"]";

    return 0;
}

File: lecture_010(RecursionwithArraylists)/l_001_getsubsequence.java
import java.util.Scanner;
import java.util.ArrayList;

public class l_001_getsubsequence {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        String str = scn.next();
        ArrayList<String> result = gss(str);
        System.out.println(result); 
    }
    //Faith: bc -> [--, -c, b-, bc]
    //Exp: abc -> [---, --c, -b-, -bc, a--, a-c, ab-, abc] 
    public static ArrayList<String> gss(String str) {
        if(str.length()==0) {
            ArrayList<String> baseres = new ArrayList<>();  //baseres = baseresult
            baseres.add("");                               
            return baseres; 
        }   
        char ch = str.charAt(0);   //a   -> 1
        String ros = str.substring(1);  //bc  -> 2          //ros = rest of string
        ArrayList<String> rres = gss(ros);   //[--, -c, b-, bc]   -> 3

        ArrayList<String> myans = new ArrayList<>();   //->4     //myans = my answer
        for(String rstr: rres) {                               
            myans.add("-" + rstr);                    //->5
        }
        for(String rstr: rres) {
            myans.add(ch + rstr);                     //->5
        }
        return myans;                               //->6
    }
}